File: solutions/typescript/113-basic-calculator.ts
/**
 * Problem 113: Basic Calculator (LeetCode 224)
 * Difficulty: Hard
 * Language: TypeScript
 */
function calculate(s: string): number {
  const chars = [...s.replace(/\s/g, '')];
  const parse = (i: number): [number, number] => {
    const go = (pos: number, result: number, sign: number): [number, number] => {
      if (pos >= chars.length || chars[pos] === ')') return [result, pos + 1];
      if (chars[pos] === '+') return go(pos + 1, result, 1);
      if (chars[pos] === '-') return go(pos + 1, result, -1);
      if (chars[pos] === '(') {
        const [val, next] = parse(pos + 1);
        return go(next, result + sign * val, 1);
      }
      const [num, next] = readNum(pos, 0);
      return go(next, result + sign * num, 1);
    };
    return go(i, 0, 1);
  };
  const readNum = (i: number, acc: number): [number, number] =>
    i < chars.length && chars[i] >= '0' && chars[i] <= '9'
      ? readNum(i + 1, acc * 10 + Number(chars[i]))
      : [acc, i];
  return parse(0)[0];
}

File: solutions/ocaml/113-basic-calculator.ml
(*
  Problem 113: Basic Calculator (LeetCode 224)
  Difficulty: Hard
  Language: OCaml
*)
let calculate s =
  let chars = String.to_seq s
    |> Seq.filter (fun c -> c <> ' ') |> List.of_seq in
  let rec parse cs = go cs 0 1
  and go cs result sign = match cs with
    | [] -> (result, [])
    | ')' :: rest -> (result, rest)
    | '+' :: rest -> go rest result 1
    | '-' :: rest -> go rest result (-1)
    | '(' :: rest ->
      let (v, rest') = parse rest in
      go rest' (result + sign * v) 1
    | _ ->
      let rec read_num cs acc = match cs with
        | c :: rest when c >= '0' && c <= '9' ->
          read_num rest (acc * 10 + Char.code c - Char.code '0')
        | _ -> (acc, cs) in
      let (num, rest) = read_num cs 0 in
      go rest (result + sign * num) 1 in
  fst (parse chars)
